libmscore/instrument.cpp

```diff
@@ -20,6 +20,8 @@
 #include "mscore.h"
 #include "part.h"
 #include "score.h"
+#include "synthesizer/msynthesizer.h"
```

Comment:
This is rather dirty practice. Now Notation Model class implementation depends on the synth class. This direct dependency prevents usage of the libmscore with other synths. Could you, please, somehow get rid of this direct dependency?

Reply:
Yep, reviewing what I wrote I can see how to fix this. Thanks for pointing it out, it slipped past me.

Reply:
Hang on, no, I'm wrong. In `Score`, which is part of libmscore, we [still use the synthesizer](https://github.com/musescore/MuseScore/blob/master/libmscore/score.cpp#L73). In fact, a score has a `SynthesizerState` member. So, I don't see any problem with these includes, it should be fine to use with `Instrument`.

Reply:
Hang on, that's not entirely correct, actually. We have synthesizer includes a few times in libmscore, intentionally or not, in [rendermidi.cpp:50](https://github.com/musescore/MuseScore/blob/master/libmscore/rendermidi.cpp#L50), [undo.h:30](https://github.com/musescore/MuseScore/blob/master/libmscore/undo.h#L30), [keyfinder.cpp:58](https://github.com/musescore/MuseScore/blob/master/libmscore/keyfinder.cpp#L58).

Reply:
As far as I understand, the main practical problem is caused not by these include statements themselves but by adding a direct dependency on `MasterSynthesizer` class and other related classes. Almost all of other includes (except the one in undo.h) are made for including `synthesizer/event.h` which contains only common classes representing MIDI events, and using these classes seems to be OK here.

Concerning switching between expressive/non-expressive patches (which seems to be the cause of introducing this dependency), it should be enough to do it only for channels registered in MIDI mapping inside MasterScore (which is, in fact, already the case as all calls to `switchExpressive` are made to `Channel`s returned by `MasterScore::playbackChannel()`). This implies the following:
- There is probably no need to traverse all instruments for this operation, hence, no need for having a member function inside `Instrument` class.
- Switching between patches does not really have to be implemented inside `MasterScore` as this operation seems to be soundfont-specific and is not related to MasterScore concept in general.

The only remaining question is whether it would be cleaner to implement the soundfont-specific logic of searching for an appropriate patch just inside `Channel` class, like it is done now, or move it to `mscore` or Synthesizer subsystem. @jthistle, what do you think on this, and don't I miss something in my review of the issue?

Reply:
Move it to mscore, I'd say. If our design principles say that we keep Q_OBJECTs out of libmscore, then let's keep it that way. Personally, I'd move the code that checks for the soundfont to the synthesizer, and then from there iterate over all the necessary channels. I'll see if I can get a PR done to do that.

Reply:
I have a possible solution to this issue. See https://github.com/mattmcclinch/MuseScore/commit/b2fe73331fccaf3735f2c73fa87abc59a592ea1a.

Reply:
See my comment on the commit